File: nte-ai-sdk-node-env-d0/nte-ai-sdk-node-env-d0/src/_llm-base-model-parser.ts
import {AzureOpenAIModelConfig} from "./azure-openai/index.js";
import {AzureLangchainJsonParsingWrapper, AzureUnderlyingRawResponse, LlmBaseModel, ModelInvokeReturn} from "./llm-base-model.js";
import {SDKException} from "./exception/index.js";

type ResponseType = "text" | "json";

interface ParsedResult {
    data?: any;
    text?: string;
    confidence?: number | null;
    prompt_tokens: number;
    completion_tokens: number;
}

const logger = {
    error: (message: string, ...args: any[]) => console.error(message, ...args),
};

class LlmBaseModelParser {
    private model: LlmBaseModel;

    constructor(model: LlmBaseModel) {
        this.model = model;
    }

    public parse(response: ModelInvokeReturn, responseType: ResponseType = "text"): ParsedResult {
        const parserFunction = this._getParser();
        return parserFunction(response, responseType);
    }

    private _getParser(): (response: ModelInvokeReturn, responseType: ResponseType) => ParsedResult {
        if (this.model.modelConfig instanceof AzureOpenAIModelConfig) {
            return this._parseModelAzureAIResult;
        }
        const errorMessage = "Unsupported model type. Please provide a valid AzureOpenAIModelConfig.";
        logger.error(errorMessage);
        throw new SDKException(400,errorMessage);
    }

    private _parseModelAzureAIResult(response: ModelInvokeReturn, responseType: ResponseType): ParsedResult {
        let underlyingRawResponse: AzureUnderlyingRawResponse;
        let parsedJsonFromWrapper: any | null = null;
        let hadJsonParsingError: boolean = false;

        if (responseType === "json") {
            const jsonWrapper = response.raw as AzureLangchainJsonParsingWrapper;
            if (typeof jsonWrapper.raw === 'undefined') {
                 logger.error("Invalid response structure for JSON parsing. Expected Langchain wrapper with 'raw' and 'parsing_error' fields.");
                 underlyingRawResponse = response.raw as AzureUnderlyingRawResponse;
                 hadJsonParsingError = jsonWrapper.parsing_error;;
            } else {
                underlyingRawResponse = jsonWrapper.raw;
                parsedJsonFromWrapper = jsonWrapper.parsed;
            }
        } else {
            underlyingRawResponse = response.raw as AzureUnderlyingRawResponse;
        }

        const metadata = underlyingRawResponse?.response_metadata;
        let confidence: number | null = null;

        if (metadata?.logprobs?.content && Array.isArray(metadata.logprobs.content)) {
            let totalLogProb = 0.0;
            for (const contentItem of metadata.logprobs.content) {
                if (typeof contentItem.logprob === 'number') {
                    totalLogProb += contentItem.logprob;
                }
            }
            if (metadata.logprobs.content.length > 0) {
                confidence = Math.exp(totalLogProb);
            }
        }

        if (responseType === "json") {
            if (!hadJsonParsingError) {
                return {
                    data: parsedJsonFromWrapper,
                    confidence: confidence,
                    prompt_tokens: response.promptTokens,
                    completion_tokens: response.completionTokens,
                };
            } else {
                return {
                    text: String(parsedJsonFromWrapper),
                    confidence: confidence,
                    prompt_tokens: response.promptTokens,
                    completion_tokens: response.completionTokens,
                };
            }
        }

        const textContent = underlyingRawResponse?.content?.trim() || "";
        return {
            text: textContent,
            confidence: confidence,
            prompt_tokens: response.promptTokens,
            completion_tokens: response.completionTokens,
        };
    }
}

export { LlmBaseModelParser, ParsedResult, ResponseType };

File: nte-aisdk-release-2.0/nte-aisdk-release-2.0/src/nte_aisdk/_llm_base_model_parser.py
from __future__ import annotations

import json
import math
import re
from typing import TYPE_CHECKING, Any, Literal

from .azure_openai import AzureOpenAIModelConfig
from .deepseek import DeepseekModelConfig
from .exception import SDKException
from .llama_ai.model_config import LlamaModelConfig
from .vertex_ai.model_config import VertexAIModelConfig

if TYPE_CHECKING:
    from .llm_base_model import LlmBaseModel

import logging

logger = logging.getLogger(__name__)

class LlmBaseModelParser:
    def __init__(self, model: LlmBaseModel):
        self.model = model

    def parse(self, response, response_type: Literal["text", "json"] = "text") -> dict[str, Any]:
        parser = self._get_parser(self.model.model_config)
        return parser(response, response_type)

    def _parse_model_azureai_result(self, response, response_type: Literal["text", "json"]) -> dict[str, Any]:
        def _parse_azure_chat_open_ai(response, response_type: Literal["text", "json"]) -> dict[str, Any]:
            raw_response = response["raw"]["raw"] if response_type == "json" else response["raw"]
            metadata = raw_response.response_metadata
            # # Retrieve metadata
            confidence = None
            if metadata.get("logprobs") is not None:
                total_log_prob = 0.0
                for content in metadata["logprobs"]["content"]:
                    total_log_prob += content["logprob"]
                confidence = math.exp(total_log_prob)

            if response_type == "json":
                # Extract the entities from the response
                if not response["raw"]["parsing_error"]:
                    return {
                        "data": response["raw"]["parsed"],
                        "confidence": confidence,
                        "prompt_tokens": response["prompt_tokens"],
                        "completion_tokens": response["completion_tokens"],
                    }
                return {
                    "text": str(response["raw"]["parsed"]),
                    "confidence": confidence,
                    "prompt_tokens": response["prompt_tokens"],
                    "completion_tokens": response["completion_tokens"],
                }

            # Extract the entities from the response
            text = response["raw"].content.strip()
            return {
                "text": text,
                "confidence": confidence,
                "prompt_tokens": response["prompt_tokens"],
                "completion_tokens": response["completion_tokens"],
            }

        # Call the inner_function with provided arguments
        return _parse_azure_chat_open_ai(response, response_type)


    def _parse_model_vertexai_result(self, response, response_type: Literal["text", "json"]) -> dict[str, Any]:
        def _parse_vertexai(response, response_type: Literal["text", "json"]) -> dict[str, Any]:
            confidence = math.exp(response["raw"].candidates[0].avg_logprobs*response["raw"].usage_metadata.candidates_token_count)
            text = response["raw"].candidates[0].content.parts[0].text
            # Extract the entities from the response
            if response_type == "json":
                try:
                    corrected_json_string = text.replace("```json", "").replace("```", "").strip()
                    json_object = json.loads(corrected_json_string)
                except json.JSONDecodeError:
                    # Handle the case where text is not a valid JSON string
                    # Keep the original value of result["text"]
                    return {
                        "text": text,
                        "confidence": confidence,
                        "prompt_tokens": response["prompt_tokens"],
                        "completion_tokens": response["completion_tokens"],
                    }
                else:
                    return {
                        "data": json_object,
                        "confidence": confidence,
                        "prompt_tokens": response["prompt_tokens"],
                        "completion_tokens": response["completion_tokens"],
                    }
            else:
                return {
                    "text": text,
                    "confidence": confidence,
                    "prompt_tokens": response["prompt_tokens"],
                    "completion_tokens": response["completion_tokens"],
                }

        # Call the inner_function with provided arguments
        return _parse_vertexai(response, response_type)

    def _parse_model_llama_result(self, response, response_type: Literal["text", "json"]) -> dict[str, Any]:
        def _parse_llama(response, response_type: Literal["text", "json"]) -> dict[str, Any]:
            text = response["raw"].choices[0].message.content
            if response_type == "json":
                try:
                    corrected_json_string = text.strip()
                    json_object = json.loads(corrected_json_string)
                except json.JSONDecodeError:
                    # Handle the case where text is not a valid JSON string
                    # Keep the original value of result["text"]
                    return {
                        "text": text,
                        "prompt_tokens": response["prompt_tokens"],
                        "completion_tokens": response["completion_tokens"],
                    }
                else:
                    return {
                        "data": json_object,
                        "prompt_tokens": response["prompt_tokens"],
                        "completion_tokens": response["completion_tokens"],
                    }
            return {
                "text": text,
                "prompt_tokens": response["prompt_tokens"],
                "completion_tokens": response["completion_tokens"],
            }

        # Call the inner_function with provided arguments
        return _parse_llama(response, response_type)

    def _parse_model_deepseek_result(self, response, response_type: Literal["text", "json"]) -> dict[str, Any]:
        def _parse_deepseek(response, response_type: Literal["text", "json"]) -> dict[str, Any]:
            raw_response = response["raw"]

            # https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/reasoning_parsers/deepseek_r1_reasoning_parser.py
            text = raw_response.content.strip()
            think_start_token = "<think>"  # noqa: S105
            think_end_token = "</think>"  # noqa: S105
            reasoning_regex = re.compile(
                rf"{think_start_token}(.*?){think_end_token}", re.DOTALL)
            if think_end_token not in text:
                return {
                    "data": {},
                    "prompt_tokens": response["prompt_tokens"],
                    "completion_tokens": response["completion_tokens"],
                }
            else:
                # Add a start token if it's missing to keep compatibility.
                if think_start_token not in text:
                    text = f"{think_start_token}{text}"
                # Use a regex to find the reasoning content
                reasoning_content = reasoning_regex.findall(text)[0]
                end_index = len(
                    f"{think_start_token}{reasoning_content}{think_end_token}"
                )
                final_output = text[end_index:]

                if response_type == "json":
                    # Extract the entities from the response
                    if len(final_output) == 0:
                        return {
                            "reasoning": reasoning_content,
                            "data": {},
                            "prompt_tokens": response["prompt_tokens"],
                            "completion_tokens": response["completion_tokens"],
                        }

                    return {
                        "reasoning": reasoning_content,
                        "data": json.loads(final_output),
                        "prompt_tokens": response["prompt_tokens"],
                        "completion_tokens": response["completion_tokens"],
                    }

                # Extract the entities from the response
                return {
                    "text": final_output,
                    "reasoning": reasoning_content,
                    "prompt_tokens": response["prompt_tokens"],
                    "completion_tokens": response["completion_tokens"],
                }

        # Call the inner_function with provided arguments
        return _parse_deepseek(response, response_type)

    def _get_parser(self, llm_model):
        if isinstance(llm_model, AzureOpenAIModelConfig):
            return self._parse_model_azureai_result
        if isinstance(llm_model, VertexAIModelConfig):
            return self._parse_model_vertexai_result
        if isinstance(llm_model, LlamaModelConfig):
            return self._parse_model_llama_result
        if isinstance(llm_model, DeepseekModelConfig):
            return self._parse_model_deepseek_result
        logger.error("Unsupported model type. Please provide a valid model type - GPT, VertexAI or Llama model")
        msg = "Unsupported model type. Please provide a valid model type - GPT, VertexAI or Llama model"
        raise SDKException(msg)
